mountpoint-s3/src/fs.rs

```diff
@@ -547,7 +548,8 @@ where
                     }
                 };
 
-                handle.inode.finish_writing(size)?;
+                let parent_lookup = self.superblock.getattr(&self.client, handle.inode.parent()).await?;
+                handle.inode.finish_writing(size, &parent_lookup.inode)?;
```

Comment:
It might be a little cleaner to push the parent lookup into `finish_writing`, maybe taking a ref to the superblock as input instead.

Or we could give every inode a pointer back to its superblock, but that creates an annoying cycle that means the superblock can never be freed (superblock points to inode, which points back to superblock).

Another option would be to move these state transition methods onto `Superblock`.

Reply:
What if we have a new type of handle to control state of the write, let's say `WriteHandle`, it can be used the same way with `ReaddirHandle` and could be extended when we support multipart upload. Does that make sense?

Reply:
Well, we already have a write handle in the `S3Filesystem`, and I assume that's where the MPU stuff will live, since the inode doesn't care about the contents of the file. We could add a separate token type to encode the state machine that `start_writing`/`finish_writing` represents. But the problem I'm trying to solve here is that now those transitions need a reference to the superblock since they need to do a lookup. I'd rather the filesystem didn't do that lookup itself, because it leaks an implementation detail of the inode (that it needs to know the parent inode when making write transitions).

Reply:
By write the handle we already have, do you mean `FileHandle`? I think it's the same problem because we don't have a reference from file handles to the superblock. But, for directories, I see that we have `DirHandle` and then there is [`ReaddirHandle`](https://github.com/awslabs/mountpoint-s3/blob/286541e395d5dd774fd4de70644b5eaa20c5ec5f/mountpoint-s3/src/inode.rs#L509-L519) inside, which holds the reference to the superblock to do readdir stuff.

Reply:
That DirHandle -> ReaddirHandle indirection is because the DirHandle (which is a filesystem concept) shouldn't know about the implementation details of the ReaddirHandle (which is an inode concept). It's kind of the same thing here: the FileHandleType (which is a filesystem concept) shouldn't know about the implementation details of the inode. So, yeah, I suppose a new WriteHandle that holds a reference to the superblock would make sense here.